R/read-write.R

```diff
@@ -0,0 +1,465 @@
+    select("child_table" = table, fks) %>%
+    unnest(fks) %>%
+    rename(parent_table = table, fk_col = column) %>%
+    # key number is needed in case of compound keys
+    # FIXME: referenced column of PK then needs to be given too
+    mutate(key_nr = row_number()) %>%
+    unnest(fk_col)
+
+  list(
+    info_tibble = info_tibble,
+    col_class_tibble = col_class_tibble,
+    pk_tibble = pk_tibble,
+    fk_tibble = fk_tibble
+  )
+}
+
+dm_write_csv_impl <- function(dm, csv_directory, zip) {
+  csv_tables <- prepare_tbls_for_def_and_class(dm, csv = TRUE)
+
+  if (!dir.exists(csv_directory)) {
+    dir.create(csv_directory)
+  }
+
+  csv_info_filename <- file.path(csv_directory, "___info_file_dm.csv")
+  csv_coltypes_filename <- file.path(csv_directory, "___coltypes_file_dm.csv")
+  csv_pk_filename <- file.path(csv_directory, "___pk_file_dm.csv")
+  csv_fk_filename <- file.path(csv_directory, "___fk_file_dm.csv")
+  csv_table_filenames <- file.path(csv_directory, paste0(src_tbls(dm), ".csv"))
+
+  readr::write_csv(csv_tables$info_tibble, csv_info_filename)
+  readr::write_csv(csv_tables$col_class_tibble, csv_coltypes_filename)
+  readr::write_csv(csv_tables$pk_tibble, csv_pk_filename)
+  readr::write_csv(csv_tables$fk_tibble, csv_fk_filename)
+
+  dm_tables <- dm_get_tables_impl(dm) %>%
+    convert_all_times_to_utc(csv_tables$col_class_tibble)
+  walk2(dm_tables, csv_table_filenames, readr::write_csv)
+  if (!zip) {
+    message(
+      glue::glue("Written `dm` as collection of csv-files to directory {tick(csv_directory)}.")
+    )
+  }
+  invisible(dm)
+}
+
+dm_col_class <- function(dm) {
+  tables <- dm_get_tables_impl(dm)
+  map(dm_get_tables_impl(dm), colnames) %>%
+    imap_dfr(function(col, table) {tibble(table = table, column = col)}) %>%
+    mutate(class = map2_chr(
+      table,
+      column,
+      function(table, column) {
+        tables[[table]] %>%
+          pull(column) %>%
+          class() %>%
+          pluck(1)}
+      )
+    )
+}
+
+readr_translate <- function(r_class) {
+  map_chr(r_class, ~ filter(class_translation_r_readr, r_class == .x) %>% pull(readr_code))
+}
+
+class_translation_r_readr <- tribble(
+  ~r_class, ~readr_code,
+  "logical", "l",
+  "integer", "i",
+  "numeric", "d",
+  "character", "c",
+  "Date", "D",
+  "POSIXlt", "T",
+  "POSIXct", "T"
+)
+
+xlsx_classes <- c(
+  "logical",
+  "integer",
+  "numeric",
+  "character",
+  "Date",
+  "POSIXlt",
+  "POSIXct"
+)
+
+convert_all_times_to_utc <- function(table_list, col_class_table) {
+  if (any(col_class_table$class %in% c("POSIXlt", "POSIXct"))) {
+    to_convert <- filter(col_class_table, class %in% c("POSIXlt", "POSIXct"))
+    message(
+      c("Converting the datetime values for the following column(s) to timezone `UTC`:\n",
+        glue::glue("{paste0(to_convert$table, '$', to_convert$column, collapse = '\n')}"))
+    )
+    table_list <- reduce2(to_convert$table, to_convert$column, function(tables, table, column) {
```

Comment:
`mutate(across())` ?

Reply:
maybe, but my thought was that it's good to inform users which columns are converted. And since then we know those columns, we can make those changes as well explicitly.